Feature/example test splcomposites

chanskw opened the pull request:
@ddebrunner This is the first set of example for testing SPL composites.

Please check this out.  We can add more documentation... but want some feedback first.

Thanks!

chanskw commented:
https://github.com/IBMStreams/streamsx.testing/tree/feature/example_test_splcomposites/examples/spl_composites

ddebrunner commented:
Just looking at the namespaces I don't think we should use `com.ibm.streamsx.testing` for any code that is examples/samples etc.

Maybe it's ok if we have a common prefix, so that say all examples toolkits were under `com.ibm.streamsx.testing.examples`, e.g. `com.ibm.streamsx.testing.examples.app1`.

But clearly separate out any examples from code and namespaces/packages that might potentially be in an SPL toolkit/Java library/Python package that provides generic testing capabilities.

ddebrunner commented:
I also wonder if initially we have a single SPL example toolkit that contains all the SPL artifacts we want to show how to test, rather than potentially many toolkits (spl composites, spl functions, java funtions, java primtives, C++ primitives, python primitives, ...)

ddebrunner commented on examples/spl_composites/com.ibm.streamsx.testing.java/src/com/ibm/streamsx/testing/java/TestCustom.java:
```diff
@@ -0,0 +1,85 @@
+ * This class hows how one can test a SPL composite operator using JUnit
+ * and the Streams Java Application API.
+ */
+public class TestCustom {
```
should lead by example on naming, this is testing the `MultipyBy` operator, not 'Custom', so rename to `TestMultipyBy`.

chanskw replied:
Done

ddebrunner commented on examples/spl_composites/com.ibm.streamsx.testing.python/testCustom.py:
```diff
@@ -0,0 +1,57 @@
+import numpy as np
+
+# TestCustomStandalone class for driving the test in standalone mode
+class TestCustomStandalone(unittest.TestCase):
```
should lead by example on naming, this is testing the MultipyBy operator, not 'Custom', so rename to `TestMultipyByStandalone`.

chanskw replied:
done

ddebrunner commented on examples/spl_composites/com.ibm.streamsx.testing.python/testCustom.py:
```diff
@@ -0,0 +1,57 @@
+        tester.test(self.test_ctxtype, self.test_config)
+
+
+# Example to run the same test in distributed mode
```
Have the comment as the doc comment:
```
class TestCustomDistributed(TestCustomStandalone):
    """  Example to run the same test in distributed mode. """
```

chanskw replied:
done

ddebrunner commented on examples/spl_composites/com.ibm.streamsx.testing.java/src/com/ibm/streamsx/testing/java/TestCustom.java:
```diff
@@ -0,0 +1,85 @@
+		StreamSchema schema = Type.Factory.getStreamSchema("tuple<int32 result>");
+		
+		// Set up parameters for the composite operator
+		Map<String, Object> params = new HashMap<String, Object>();
```
No need to have types on RHS: `new HashMap<`~String, Object~`>()`

chanskw replied:
I am keeping this... as not doing it generates a warning in Eclipse.  Trying to minimize any warning.

ddebrunner commented on examples/spl_composites/com.ibm.streamsx.testing.java/src/com/ibm/streamsx/testing/java/TestCustom.java:
```diff
@@ -0,0 +1,85 @@
+		// Invoke the test composite
+		SPLStream resultStream = SPL.invokeSource(topo, "com.ibm.streamsx.testing.spl::TestMultiplyBy", params, schema);
+		
+		// Convert the SPL Stream to TStream so we can work with it with the tester
```
In this case it's not clear why you are converting the SPL stream to a TStream since you are only checking the count.

chanskw replied:
I was setting up for us to be able to validate the stream content.  I did the count condition because it was easiest to do.

ddebrunner commented on examples/spl_composites/com.ibm.streamsx.testing.python/testCustom.py:
```diff
@@ -0,0 +1,57 @@
+        ''' Call the test composite'''
+        testStream = op.Source(topo, 'com.ibm.streamsx.testing.spl::TestMultiplyBy', 'tuple<int32 result>', params=params)
+
+        ''' Convert the SPLStream to Python Stream so we can work with the data in the tester '''
```
The tester can handle SPL/structured stream so it's not clear why you are converting here.

chanskw replied:
Can you show how?  I couldn't get it to work  without going down this path.

chanskw commented:
Can you elaborate on how one would test MultiplBy without a wrapper composite?  Are you suggesting that we generate data in toopology (java / python) and sending it in for testing?

I took that path before and it was more complicated as user has to understand how to generate data in the test harness, feed the data to the composite (crossing language boundary), and then coming back to do validation.  This was simpler and more natural, using a beacon was much simpler than generating the same thing in java and python.

chanskw commented:
I can make the changes for the namespace.. perhaps it's easier if I make those changes first and merge it in, and you can modify the tests itself?

chanskw commented:
I also thought about having a big toolkit project.. and we show how to test them all.
I do not know if that's going to make the project so complicated that it makes it hard to understand the example.  That's why I started smaller... trying to have very targeted examples.

chanskw commented:
I struggled with the naming of the projects too.  Let me try out your suggesions

chanskw commented:
I have addressed most of your comments.  I am merging this one in and we can continue to improve this.  I think we need some documentation / article to explain this.
